client/src/http.rs

```diff
@@ -1,35 +1,337 @@
+    /// describe how to use this module from both **flow implemention** (data side) and
+    /// **transport implementation** sides.
+    ///
+    /// ## Flow implementation
+    ///
+    /// From data side, you should implement a state machine built on top of these traits:
+    ///
+    /// - [`Init`] - it is designed to consume it's impl struct and produce a tuple, that has 2 items:
+    ///   **initial data** to establish WS connection, and the **handler** of the next flow stage - **handshake**.
+    ///   Then, transportation side should open a connection, send first message into it, receive message from Iroha
+    ///   and pass it into the next handler.
+    /// - [`Handshake`] - handles incoming message and ensures that it is OK. Which message is OK -
+    ///   implementation dependent. If it is OK, returns the handler for the next, final flow stage - **events**.
+    /// - [`Events`] - handles incoming messages and returns a **binary reply** back Iroha and **some decoded event**.
+    ///
+    /// Here is an example of how to implement flow in a transport-agnostic manner:
+    ///
+    /// ```rust
+    /// use eyre::{eyre, Result};
+    /// use iroha_client::http::{
+    ///     ws::conn_flow::{
+    ///         EventData, Events as FlowEvents, Handshake as FlowHandshake, Init as FlowInit,
+    ///         InitData,
+    ///     },
+    ///     Method, RequestBuilder,
+    /// };
+    ///
+    /// struct Init;
+    ///
+    /// impl<R: RequestBuilder> FlowInit<R> for Init {
+    ///     type Next = Handshake;
+    ///
+    ///     fn init(self) -> InitData<R, Self::Next> {
+    ///         InitData::new(
+    ///             R::new(Method::GET, "http://localhost:3000"),
+    ///             vec![1, 2, 3],
+    ///             Handshake,
+    ///         )
+    ///     }
+    /// }
+    ///
+    /// struct Handshake;
+    ///
+    /// impl FlowHandshake for Handshake {
+    ///     type Next = Events;
+    ///
+    ///     fn message(self, message: Vec<u8>) -> Result<Self::Next> {
+    ///         if message[0] == 42 {
+    ///             Ok(Events)
+    ///         } else {
+    ///             Err(eyre!("Wrong"))
+    ///         }
+    ///     }
+    /// }
+    ///
+    /// struct Events;
+    ///
+    /// impl FlowEvents for Events {
+    ///     type Event = u8;
+    ///
+    ///     fn message(&self, message: Vec<u8>) -> Result<EventData<Self::Event>> {
+    ///         Ok(EventData::new(message[0], vec![3, 2, 1]))
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// ## Transport implementation
+    ///
+    /// You are a library user and want to use Iroha Client with your own HTTP/WS implementation. For such a purpose
+    /// the client library should provide an API wrapped into the flow traits. Anyway, firstly you should implement
+    /// [`super::RequestBuilder`] trait for your transport.
+    ///
+    /// Let's take Events API as an example. [`crate::client::Client::events_handler()`] function creates a struct of
```

Comment:
Fixed. Btw it is not a mistake to put parentheses... But here and in other places they are extra, I agree.

Reply:
I think `events_handler` can be brought into scope, and we should probably use an unqualified name in docs (since all docs are hyperlinked by default). The parentheses disambiguate in cases where you have shadowing. Typically only a problem with traits, macros and structs, but occasionally we get a module named after a function, so I don't mind keeping the parens.